executor/write.go

```diff
@@ -65,50 +70,50 @@ func updateRecord(ctx context.Context, h int64, oldData, newData []types.Datum,
 			}
 			t.RebaseAutoID(val, true)
 		}
-		casted, err := table.CastValue(ctx, newData[i], col.ToInfo())
+		cmp, err := newData[i].CompareDatum(sc, oldData[i])
 		if err != nil {
 			return false, errors.Trace(err)
 		}
-		newData[i] = casted
-		touched[i] = true
-		assignExists = true
-	}
-
-	// If no assign list for this table, no need to update.
-	if !assignExists {
-		return false, nil
+		if cmp != 0 {
+			changed = true
+			touched[i] = true
+			if col.IsPKHandleColumn(t.Meta()) {
+				handleChanged = true
+			}
+		} else {
+			if mysql.HasOnUpdateNowFlag(col.Flag) && touched[i] {
+				// It's for "UPDATE t SET ts = ts" and ts is a timestamp.
+				onUpdateNoChange[i] = true
+			}
+			touched[i] = false
+		}
 	}
 
-	if err := table.CheckNotNull(cols, newData); err != nil {
+	err := table.CheckNotNull(t.Cols(), newData)
+	if err != nil {
 		return false, errors.Trace(err)
 	}
 
-	// If row is not changed, we should do nothing.
-	rowChanged := false
-	for i := range oldData {
-		if !touched[i] {
-			continue
-		}
-
-		n, err := newData[i].CompareDatum(sc, oldData[i])
-		if err != nil {
-			return false, errors.Trace(err)
-		}
-		if n != 0 {
-			rowChanged = true
-			break
-		}
-	}
-	if !rowChanged {
+	if !changed {
 		// See https://dev.mysql.com/doc/refman/5.7/en/mysql-real-connect.html  CLIENT_FOUND_ROWS
 		if ctx.GetSessionVars().ClientCapability&mysql.ClientFoundRows > 0 {
 			sc.AddAffectedRows(1)
 		}
 		return false, nil
 	}
 
-	var err error
-	if !newHandle.IsNull() {
+	// Only if some fields are really changed, we can change those on-update fields.
+	for i, col := range t.Cols() {
+		if mysql.HasOnUpdateNowFlag(col.Flag) && !touched[i] && !onUpdateNoChange[i] {
```

Comment:
for this case:
```sql
update t set ts_field = ts_field;
```
the `ts_field` has attributes `timestamp not null default current_timestamp on update current_timestamp`.

Suppose the `offset` of `ts_field` is 3, so `touched[3]` is `false`, because it's not changed. But we can't set it to `current_timestamp`, because we specied that field explicitly in `set list`.